elements/howto-checkbox/README.md

```diff
@@ -1,5 +1,180 @@
+its own values.
+
+### Make properties lazy
+
+A developer may attempt to set a property on the element before its definition
+has been loaded. This is especially true if the developer is using a framework
+which handles loading components, stamping them to the page, and binding their
+properties to a model.
+
+```html
+<!--
+Here Angular is declaratively binding its model's isChecked property to the
+checkbox's checked property.
+-->
+<howto-checkbox [checked]="defaults.isChecked"></howto-checkbox>
+```
+
+A Custom Element should be able to handle this scenario by checking its instance
+for any already set properties during its connected phase. The
+`<howto-checkbox>` demonstrates this pattern using a method called
+`_upgradeProperty`.
+
+```js
+connectedCallback() {
+  ...
+  this._upgradeProperty('checked');
+}
+
+_upgradeProperty(prop) {
+  if (this.hasOwnProperty(prop)) {
+    let value = this[prop];
+    delete this[prop];
+    this[prop] = value;
+  }
+}
+```
+
+`_upgradeProperty` captures the value from the unupgraded instance and deletes
+the property so it does not shadow the Custom Element's own property setter.
+This way, when the element's definition does finally load, it can immediately
+reflect the correct state.
+
+### Avoid reentrancy issues
+
+It's tempting to use the `attributeChangedCallback` to reflect state to an
+underlying property, for example:
+
+```js
+// When the [checked] attribute changes, set the checked property to match.
+attributeChangedCallback(name, oldValue, newValue) {
+  if (name === 'checked')
+    this.checked = newValue;
+}
+```
+
+But this can create an infinite loop if the property setter also reflects to
+the attribute.
+
+```js
+set checked(value) {
+  const isChecked = Boolean(value);
+  if (isChecked)
+    // OOPS! This will cause an infinite loop because it triggers the
+    // attributeChangedCallback() which then sets this property again.
+    this.setAttribute('checked', '');
+  else
+    this.removeAttribute('checked');
+}
+```
+
+An alternative is to allow the property setter to reflect to the attribute, and
+have the getter determine it's value based on the attribute.
+
+```js
+set checked(value) {
+  const isChecked = Boolean(value);
+  if (isChecked)
+    this.setAttribute('checked', '');
+  else
+    this.removeAttribute('checked');
+}
+
+get checked() {
+  return this.hasAttribute('checked');
+}
+```
+
+In this example, removing the attribute will also effectively set the property
+to `false`.
+
+Finally, the `attributeChangedCallback` can be used to handle side effects like
+applying ARIA states.
+
+```js
+attributeChangedCallback(name, oldValue, newValue) {
+  const hasValue = newValue !== null;
+  switch (name) {
+    case 'checked':
+      // Note the attributeChangedCallback is only handling the *side effects*
+      // of setting the attribute.
+      this.setAttribute('aria-checked', hasValue);
+      break;
+    ...
+  }
+}
+```
+
+
+## Reference
```

Comment:
I was thinking it's better singular, as in "the reference section". Same as how the previous section is "summary". I can listify the links.